src/device.hpp

```diff
@@ -300,10 +300,31 @@ struct cvk_device : public _cl_device_id,
     bool supports_subgroups() const { return m_has_subgroups_support; }
 
     bool supports_non_uniform_decoration() const {
-        return m_properties.apiVersion >= VK_MAKE_VERSION(1, 2, 0) ||
-               is_vulkan_extension_enabled(
-                   VK_EXT_DESCRIPTOR_INDEXING_EXTENSION_NAME);
+        bool supported = m_properties.apiVersion >= VK_MAKE_VERSION(1, 2, 0) ||
+                         is_vulkan_extension_enabled(
+                             VK_EXT_DESCRIPTOR_INDEXING_EXTENSION_NAME);
+
+        // This is a hack to work-around a driver bug in mali device using the
+        // bifrost driver.
+        if (m_clvk_properties->get_non_uniform_decoration_unsupported()) {
+            return false;
+        }
+
+        if (supported && !(m_features_descriptor_indexing
+                               .shaderSampledImageArrayNonUniformIndexing &&
+                           m_features_descriptor_indexing
+                               .shaderStorageImageArrayNonUniformIndexing &&
+                           m_features_descriptor_indexing
+                               .shaderUniformBufferArrayNonUniformIndexing &&
+                           m_features_descriptor_indexing
+                               .shaderStorageBufferArrayNonUniformIndexing)) {
+            cvk_warn_fn("Device should support non-uniform decorations, but "
+                        "some expected features are set to false. This is "
+                        "unexpected and might lead to errors");
+        }
```

Comment:
```suggestion
```

It's perfectly valid for an implementation to support non-uniform indexing for only some descriptor types or even none but to correctly support the non-uniform decoration. Better to remove this IMO.